actors/market/src/lib.rs

```diff
@@ -683,10 +780,18 @@ impl Actor {
         let miner_addr = rt.message().caller();
 
         rt.transaction(|st: &mut State, rt| {
-            let mut deal_states: Vec<(DealID, DealState)> = vec![];
+            // The sector deals mapping is removed all at once.
+            // Other deal clean-up is deferred to cron.
+            let sector_deals =
+                st.pop_sector_deal_ids(rt.store(), &miner_addr, params.sectors.iter())?;
+            let all_deal_ids = sector_deals
+                .iter()
+                .flat_map(|(_, deal_ids)| deal_ids.deals.iter())
+                .collect::<Vec<_>>();
 
-            for id in params.deal_ids {
-                let deal = st.find_proposal(rt.store(), id)?;
+            let mut deal_states: Vec<(DealID, DealState)> = vec![];
+            for id in all_deal_ids {
+                let deal = st.find_proposal(rt.store(), *id)?;
```

Comment:
I know you didn't change this but.... AHHHHHHHHHH, that's loading the proposals array for every deal ID _inside cron_.